src/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,222 @@
+        | otherwise
+            = pred pageCount
+
+    protoStart :: Int
+    protoStart = searchForProtoStart 0 (length lastKeys)
+
+    searchForProtoStart :: Int -> Int -> Int
+    searchForProtoStart first count
+        | count == 0               = first
+        | key <= lastKeys ! center = searchForProtoStart first frontCount
+        | otherwise                = searchForProtoStart (succ center) rearCount
+        where
+
+        frontCount, rearCount, center :: Int
+        frontCount = count `div` 2
+        rearCount  = pred (count - frontCount)
+        center     = first + frontCount
+
+    pageCount :: Int
+    pageCount = length lastKeys
+
+-- * Deserialisation
+
+{-|
+    Reads an index along with the number of entries of the respective run from a
+    byte string.
+
+    The byte string must contain the serialised index exactly, with no leading
+    or trailing space.
+-}
+fromSBS :: ShortByteString -> Either String (NumEntries, IndexOrdinary)
+fromSBS shortByteString@(ShortByteString byteArray)
+    | fullSize < 12
+        = Left "Doesn't contain header and footer"
+    | version == byteSwap32 supportedVersion
+        = Left "Non-matching endianness"
+    | version /= supportedVersion
+        = Left "Unsupported version"
+    | otherwise
+        = (,) <$> entriesCount <*> index
+    where
+
+    fullSize :: Int
+    fullSize = ShortByteString.length shortByteString
+
+    fullBytes :: Primitive.Vector Word8
+    fullBytes = Primitive.Vector 0 fullSize byteArray
+
+    version :: Word32
+    version = indexByteArray byteArray 0
+
+    postVersionBytes :: Primitive.Vector Word8
+    postVersionBytes = Primitive.drop 4 fullBytes
+
+    lastKeysBytes, entriesCountBytes :: Primitive.Vector Word8
+    (lastKeysBytes, entriesCountBytes)
+        = Primitive.splitAt (fullSize - 12) postVersionBytes
+
+    entriesCount :: Either String NumEntries
+    entriesCount
+        | (fromIntegral asWord64 :: Integer) > fromIntegral (maxBound :: Int)
+            = Left "Number of entries not representable as Int"
+        | otherwise
+            = Right (NumEntries (fromIntegral asWord64))
+        where
+
+        asWord64 :: Word64
+        asWord64 = indexByteArray entriesCountRep 0
+
+        entriesCountRep :: ByteArray
+        Primitive.Vector _ _ entriesCountRep = Primitive.force entriesCountBytes
+
+    index :: Either String IndexOrdinary
+    index = IndexOrdinary <$> fromList <$> lastKeys lastKeysBytes
+
+    lastKeys :: Primitive.Vector Word8 -> Either String [SerialisedKey]
+    lastKeys bytes
+        | Primitive.null bytes
+            = Right []
+        | otherwise
+            = do
+                  when (Primitive.length bytes < 2)
+                       (Left "Too few bytes for key size")
+                  let
+
+                      firstSizeRep :: ByteArray
+                      Primitive.Vector _ _ firstSizeRep
+                          = Primitive.force (Primitive.take 2 bytes)
```

Comment:
This `force` is not necessary

Reply:
I think it is, for the same reasons as in the case discussed [above](https://github.com/IntersectMBO/lsm-tree/pull/244#discussion_r1638534358).

Reply:
`firstSizeRep` is here only used in `firstSize`, and it is not kept alive after that

Reply:
And also my [above follow-up comment](https://github.com/IntersectMBO/lsm-tree/pull/244#discussion_r1644652360) applies here, I think. 🙂